src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs

```diff
@@ -18,82 +18,23 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
-            internal bool IsDefault => Volatile.Read(ref _extensions) == null;
-
-            /// <summary>
-            /// Set the fields of the builder.
-            /// </summary>
-            /// <remarks>
-            /// This method guarantees: fields will be set once; exactly one caller is
-            /// returned true; and IsDefault will return true until all fields are initialized.
-            /// This method does not guarantee that all fields will be set by the same
-            /// caller. Instead, the expectation is that all callers will attempt to initialize
-            /// the builder with equivalent TypeWithAnnotations instances where
-            /// different fields of the builder may be assigned from different instances.
-            /// </remarks>
-            internal bool InterlockedInitialize(TypeWithAnnotations type)
-            {
-                if (!IsDefault)
-                {
-                    return false;
-                }
-
-                Interlocked.CompareExchange(ref _nullableAnnotation, (int)type.NullableAnnotation, 0);
-                Interlocked.CompareExchange(ref _defaultType, type.DefaultType, null);
-
-                // Because _extensions always gets a non-null value when the struct is initialized, we use it
-                // as a flag to signal that the initialization is complete. This means _extensions should be
-                // initialized last.
-                bool wasFirst = Interlocked.CompareExchange(ref _extensions, type._extensions ?? Extensions.Default, null) is null;
-
-                Debug.Assert(_extensions == (type._extensions ?? Extensions.Default));
-                Debug.Assert(_nullableAnnotation == (int)type.NullableAnnotation);
-                Debug.Assert(_defaultType.Equals(type.DefaultType));
-                return wasFirst;
-            }
+            internal readonly TypeWithAnnotations Value;
 
-            /// <summary>
-            /// We should not be adding any new usages of this method. The only one is currently in SourcePropertySymbol,
-            /// which currently sets the property's type twice in error scenarios.
-            /// We should be able to remove this method by fixing https://github.com/dotnet/roslyn/issues/35381
-            /// </summary>
-            [EditorBrowsable(EditorBrowsableState.Never)]
-            internal void InterlockedDangerousReset()
+            private Boxed(TypeWithAnnotations value)
             {
-                Interlocked.Exchange(ref _nullableAnnotation, 0);
-                Interlocked.Exchange(ref _defaultType, null);
-                Interlocked.Exchange(ref _extensions, null);
+                Value = value;
             }
 
-            /// <summary>
-            /// Create immutable TypeWithAnnotations instance.
-            /// </summary>
-            internal TypeWithAnnotations ToType()
+            internal static Boxed InterlockedCompareExchange(ref Boxed target, TypeWithAnnotations value)
```

Comment:
>internal static Boxed InterlockedCompareExchange(ref Boxed target, TypeWithAnnotations value) [](start = 12, length = 93)

I think that having this method is not helping anyone. Consumers decide if and how to deal with possible threading issues. This object is a readonly box for a structure

Reply:
In addition having this method doesn't prevent anyone from still initializing fields with pure assignments

---
In reply to: [280913749](https://github.com/dotnet/roslyn/pull/35459#discussion_r280913749) [](ancestors = 280913749)

Reply:
I've reverted that change to move forward and because it did have some arguably negative effect on property ctor. But I want to clarify that the goal was not to *prevent* certain usages (as in 100%), as to encourage proper usage. This is common practice in API design.

---
In reply to: [280916283](https://github.com/dotnet/roslyn/pull/35459#discussion_r280916283) [](ancestors = 280916283,280913749)

Reply:
> I've reverted that change to move forward and because it did have some arguably negative effect on property ctor. But I want to clarify that the goal was not to prevent certain usages (as in 100%), as to encourage proper usage. This is common practice in API design.

Thank you for reverting the change. The pattern was preventing reasonable and otherwise safe usage scenarios.

---
In reply to: [280923336](https://github.com/dotnet/roslyn/pull/35459#discussion_r280923336) [](ancestors = 280923336,280916283,280913749)